### ISSUE TITLE
[HELP] How can I create a scrollable panel containing custom controls?

### ISSUE THREAD

**Issue by Genhis** [nifty-gui/nifty-gui#449]
I want to create a panel with categories, where every category contains a series of images. I created custom control to lay out the images using `childLayout="absolute-inside"` and computing their position every time an image is added or removed. Then, I put the control with the corresponding label into a panel with `childLayout="vertical"`. It works well until the main panel runs out of space.

![niftygui](https://user-images.githubusercontent.com/11155942/35773316-3ac5b6b0-0946-11e8-9166-434e484f26ae.png)

I would like to use a scrollbar. I looked at `ListBox` implementation and saw that it uses the scrollbar only for changing content of pre-created labels. Therefore, I realized that `Scrollbar` probably can't be used in my case without adding too much logic.

Then, I tried `ScrollPanel` but it messed a lot with my control definitions and I ended up with incorrect parameters for `width` and `height` (probably has something to do with styles). Maybe I just don't know how to use it correctly but when I tried, no content was displayed inside the panel. The other thing that discourages me from using `ScrollPanel` is the problem with styles (issue #105 ).

Any ideas how this could be implemented? (see attached picture)

___

The problem is that, I don't know how to display only part of the panel's content and when I use absolute layout, the elements are overlapping other controls.

Here is my control definition and assigned styles:
```
<controlDefinition name="itemselector" style="itemselector" childRootId="#content" controller="itemselector.ItemSelectorController">
	<panel style="#content" childLayout="absolute-inside">
		<image style="#item" /> <!-- TEMPLATE -->
	</panel>
</controlDefinition>
```
```
<style id="itemselector#item">
	<attributes width="64px" height="64px" />
</style>
```

In the controller, I then get the width of the panel and the template image. However, they are either both zero or I'm unable to find the elements when I used `ScrollPanel`.
```
int pWidth = this.getElement().getWidth(); //panel
int iWidth = this.getElement().getChildren().get(0).getWidth(); //template
```

**Comment by Genhis** [comment]
@void256 Is the library still in development?

**Comment by bgroenks96** [comment]
@Genhis The library is in maintenance mode, so we're not actively developing new features, just fixing bugs. This smells like a bug, though. Nothing you're doing jumps out to me as super wrong. Can you post some screenshots of what you're seeing as well as post an SSCE on the [Nifty issue repository](https://github.com/nifty-gui/nifty-gui-issues)?

Thanks.

**Comment by Genhis** [comment]
@bgroenks96 I found two issues:

1. Adding a new element to ScrollPanel doesn't update the size of its page/scrollbar
2. Adding a panel to ScrollPanel with layout absolute-inside invalidates layout type of its parent. Example below:

XML structure:
```
<control id="myScrollPanel" name="scrollPanel" width="400" height="400">
	<panel id="cpanel" childLayout="vertical" backgroundColor="#0f0f" />
</control>
```
Adding elements in Java:
```
Element parent = nifty.findElementById("cpanel");

for(int i=0; i < 5; i++) {
	TextBuilder text = new TextBuilder();
	text.text("Text");
	text.style("base-font");
	ext.build(nifty, screen, parent);

	PanelBuilder panel = new PanelBuilder();
	panel.childLayout(ChildLayoutType.AbsoluteInside);
	panel.build(nifty, screen, parent);
}
```
All the text elements are placed on top of each other despite the vertical layout of the parent panel.

Do you want me to upload examples for the issues? Separately?

**Comment by bgroenks96** [comment]
Yes, to the issue repository. That will allow me to directly reproduce what you're seeing.

**Comment by Genhis** [comment]
@bgroenks96 Done.

**Comment by Genhis** [comment]
I fixed the issues by calling `layoutElements()` to ScrollPanel element and setting `width` and `height` constraints for my inner panel (recomputing every time an element is added or removed).

But I think `ScrollPanel` should do it internally when its children are modified - at least call `layoutElements()` method.

**Comment by bgroenks96** [comment]
Hi,
I apologize for not responding sooner. I was unfortunately very busy and completely forgot to follow up on your issues. In the future, if you don't get a response within a few days, please feel free to post a comment mentioning me or @void256 to get our attention.

I reviewed your SSCE for this issue and, after fixing some typos in the XMLs, was able to confirm your observations.

Could you clarify what your workaround was and where you call `layoutElements` on the `ScrollPanel`'s child root element (maybe post a code snippet)?

I could possibly implement something similar to what you suggested as a fix in `ScrollPanelControl`, but I will need to do more testing to verify that it doesn't have any side effects. There are also performance implications since we would then be performing a full relayout on the `ScrollPanelControl` after each update to a child; that could get costly in cases where the child root element has many descendants.

**Comment by Genhis** [comment]
First of all, you need to update the `height` of the parent when you add a new element. Then, I found out that it is enough to call `ScrollPanel#layoutCallback()` and full relayout is not necessary.

I don't know if Nifty can set the correct `height` implicitly (it would be much easier with that).

Here is the modified code of the **issue 1a** which I posted to `nifty-gui-issues` repository:

```
Element parent = this.screen.findElementById("cpanel");
for(int i=0; i < 50; i++) {
	//every time an element is added, parent's height has to be recomputed
	//it could be done after the element was built but another layoutElements() call would be needed
	//so it's better to know height of the element; in this case, 16
	parent.setHeight(parent.getHeight() + 16);
	
	TextBuilder text = new TextBuilder();
	text.text("Lorem ipsum dolor sit amet");
	text.style("base-font");
	text.color("#000f");
	text.build(this.nifty, this.screen, parent);
}

ScrollPanel sp = screen.findNiftyControl("sp", ScrollPanel.class);
sp.layoutCallback();
```
